jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ChunkAccumulator.java

```diff
@@ -0,0 +1,228 @@
+            }
+        };
+        return task.start();
+    }
+
+    /**
+     * @param source The {@link Content.Source} to read
+     * @param pool The {@link ByteBufferPool} to acquire the buffer from, or null for a non {@link Retainable} buffer
+     * @param direct True if the buffer should be direct.
+     * @param maxSize The maximum size to read, or -1 for no limit
+     * @return A {@link CompletableFuture} that will be completed when the complete content is read or
+     * failed if the max size is exceeded or there is a read error.
+     */
+    public CompletableFuture<RetainableByteBuffer> readAll(Content.Source source, ByteBufferPool pool, boolean direct, int maxSize)
+    {
+        CompletableTask<RetainableByteBuffer> task = new AccumulatorTask<>(source, maxSize)
+        {
+            @Override
+            protected RetainableByteBuffer take(ChunkAccumulator accumulator)
+            {
+                return accumulator.take(pool, direct);
+            }
+        };
+        return task.start();
+    }
+
+    private abstract static class AccumulatorTask<T> extends CompletableTask<T>
+    {
+        private final Content.Source _source;
+        private final ChunkAccumulator _accumulator = new ChunkAccumulator();
+        private final int _maxSize;
+
+        private AccumulatorTask(Content.Source source, int maxSize)
+        {
+            _source = source;
+            _maxSize = maxSize;
+        }
+
+        @Override
+        public void run()
+        {
+            try
+            {
+                while (true)
+                {
+                    Chunk chunk = _source.read();
+                    if (chunk == null)
+                    {
+                        _source.demand(this);
+                        return;
+                    }
+
+                    if (Chunk.isFailure(chunk))
+                    {
+                        completeExceptionally(chunk.getFailure());
+                        return;
+                    }
+
+                    if (chunk.hasRemaining())
+                    {
+                        _accumulator.add(chunk);
+
+                        if (_maxSize > 0 && _accumulator._size > _maxSize)
+                        {
+                            _accumulator.close();
+                            IOException ioe = new IOException("too large");
+                            _source.fail(ioe);
+                            completeExceptionally(ioe);
+                            return;
+                        }
+                    }
+
+                    chunk.release();
+
+                    if (chunk.isLast())
+                    {
+                        complete(take(_accumulator));
+                        return;
+                    }
+                }
+            }
+            catch (Throwable e)
+            {
+                _accumulator.close();
+                _source.fail(e);
+                completeExceptionally(e);
+            }
```

Comment:
`try/catch` should be reorganized, because if there is an exception thrown, the chunk won't be released.
Should be moved inside the `while`, I think.